google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2423,6 +2471,48 @@ private void releaseSession(PooledSession session, Position position) {
     }
   }
 
+  private boolean isUnbalanced(PooledSession session) {
+    // Don't bother with any randomization if the number of checked out sessions is low, as it is
+    // better to re-use sessions as much as possible in a low-QPS scenario.
+    if (sessions.isEmpty() || checkedOutSessions.size() <= 2) {
+      return false;
+    }
+    int numChannels = sessionClient.getSpanner().getOptions().getNumChannels();
+    if (numChannels == 1) {
+      return false;
+    }
+
+    // Ideally, the first numChannels sessions in the pool should contain exactly one session for
+    // each channel.
+    // Check if the first numChannels sessions at the head of the pool already contain more than 2
+    // sessions that use the same channel as this one. If so, we randomize.
+    int channel = session.getChannel();
+    int count = 0;
+    for (int i = 0; i < Math.min(numChannels, sessions.size()); i++) {
+      PooledSession otherSession = sessions.get(i);
+      if (channel == otherSession.getChannel()) {
+        count++;
+        if (count > 2) {
+          return true;
+        }
+      }
+    }
+    // Ideally, the use of a channel in the checked out sessions is exactly
+    // numCheckedOut / numChannels
+    // We check whether we are more than a factor two away from that perfect distribution.
```

Comment:
Benchmarking showed that this was a reasonable trade-off. There are two competing scenarios here:
1. A low-QPS scenario benefits from as little randomization as possible, because in that case it helps to reuse the same session for most queries (better cache hit ratio), while there is no real probability of overloading a single channel. Being too aggressive showed higher latencies for this scenario.
2. A high-QPS scenario benefits from as good distribution as possible, to prevent overloading a single channel.

We don't keep track of the QPS in the session pool, so the best substitute that we have for it is the number of checked out sessions. The factor 2 was chosen after trying varying factors and also other strategies, as this gave good results both for low and high QPS. (Although I've not tried with every single number between let's say 1.0 and 5.0)

Reply:
Thanks for the explanation. The reasoning sounds good.